test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/GetThreadStateMountedTest.java

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8309612 8310584
+ * @summary The test verifies that JVMTI GetThreadState function reports expected state
+ *          for mounted (pinned) virtual thread and its carrier thread
+ * @requires vm.jvmti
+ * @requires vm.continuations
+ * @run main/othervm/native
+ *      -agentlib:GetThreadStateMountedTest
+ *      GetThreadStateMountedTest
+ */
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * The test implements different scenarios to get desired JVMTI thread states.
+ * For each scenarios test also checks states after carrier and virtual threads suspend/resume
+ * and after virtual thread interrupt.
+ * Special handling is required for WAITING state scenarios:
+ * Spurious wakeups may cause unexpected thread state change and this causes test failure.
+ * To avoid this test thread should be suspended (i.e. carrier and/or mounted virtual thread is suspended).
+ */
+public class GetThreadStateMountedTest {
+
+    static final int JVMTI_THREAD_STATE_RUNNABLE                 = 0x0004;
+    static final int JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER = 0x0400;
+    static final int JVMTI_THREAD_STATE_WAITING                  = 0x0080;
+    static final int JVMTI_THREAD_STATE_WAITING_INDEFINITELY     = 0x0010;
+    static final int JVMTI_THREAD_STATE_WAITING_WITH_TIMEOUT     = 0x0020;
+    static final int JVMTI_THREAD_STATE_IN_OBJECT_WAIT           = 0x0100;
+    static final int JVMTI_THREAD_STATE_SLEEPING                 = 0x0040;
+    static final int JVMTI_THREAD_STATE_PARKED                   = 0x0200;
+    static final int JVMTI_THREAD_STATE_IN_NATIVE                = 0x400000;
+
+    static void runnable() throws Exception {
+        TestStatus status = new TestStatus("JVMTI_THREAD_STATE_RUNNABLE");
+        CountDownLatch ready = new CountDownLatch(1);
+        final boolean[] stopFlag = new boolean[1];
```

Comment:
Q: It is not clear why an array is needed instead of non-final local. The same question for `waiting()` method below,

Reply:
It's used in lambda, so must be final
I think both "flag to stop" and "flag to continue" are clear, I prefer "flag to stop" as it does not require additional initialization

Reply:
Okay, I see why an array is needed.
It is really minor and up to you. Just wanted to get rid of one negation. :)